File: subprojects/infra/src/main/java/gg/essential/cosmetics/CosmeticSlot.java
/*
 * Copyright (c) 2024 ModCore Inc. All rights reserved.
 *
 * This code is part of ModCore Inc.'s Essential Mod repository and is protected
 * under copyright registration # TX0009138511. For the full license, see:
 * https://github.com/EssentialGG/Essential/blob/main/LICENSE
 *
 * You may not use, copy, reproduce, modify, sell, license, distribute,
 * commercialize, or otherwise exploit, or create derivative works based
 * upon, this file or any other in this repository, all of which is reserved by Essential.
 */
package gg.essential.cosmetics;

import gg.essential.lib.gson.JsonDeserializationContext;
import gg.essential.lib.gson.JsonDeserializer;
import gg.essential.lib.gson.JsonElement;
import gg.essential.lib.gson.JsonParseException;
import gg.essential.lib.gson.JsonPrimitive;
import gg.essential.lib.gson.JsonSerializationContext;
import gg.essential.lib.gson.JsonSerializer;
import gg.essential.lib.gson.annotations.JsonAdapter;

import java.lang.reflect.Type;
import java.util.concurrent.ConcurrentHashMap;

@JsonAdapter(CosmeticSlot.Serializer.class)
public class CosmeticSlot {
    private static final ConcurrentHashMap<String, CosmeticSlot> ENTRIES = new ConcurrentHashMap<>();

    public final String id;

    private CosmeticSlot(String id) {
        this.id = id;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof CosmeticSlot)) return false;
        CosmeticSlot that = (CosmeticSlot) o;
        return id.equals(that.id);
    }

    @Override
    public int hashCode() {
        return id.hashCode();
    }

    @Override
    public String toString() {
        return this.id;
    }

    public static CosmeticSlot of(String id) {
        return ENTRIES.computeIfAbsent(id, CosmeticSlot::new);
    }

    public static class Serializer implements JsonSerializer<CosmeticSlot>, JsonDeserializer<CosmeticSlot> {
        @Override
        public CosmeticSlot deserialize(JsonElement jsonElement, Type type, JsonDeserializationContext jsonDeserializationContext) throws JsonParseException {
            return of(jsonElement.getAsString());
        }

        @Override
        public JsonElement serialize(CosmeticSlot cosmeticSlot, Type type, JsonSerializationContext jsonSerializationContext) {
            return new JsonPrimitive(cosmeticSlot.id);
        }
    }
}

File: cosmetics/src/commonMain/kotlin/gg/essential/mod/cosmetics/CosmeticSlot.kt
/*
 * Copyright (c) 2024 ModCore Inc. All rights reserved.
 *
 * This code is part of ModCore Inc.'s Essential Mod repository and is protected
 * under copyright registration # TX0009138511. For the full license, see:
 * https://github.com/EssentialGG/Essential/blob/main/LICENSE
 *
 * You may not use, copy, reproduce, modify, sell, license, distribute,
 * commercialize, or otherwise exploit, or create derivative works based
 * upon, this file or any other in this repository, all of which is reserved by Essential.
 */
package gg.essential.mod.cosmetics

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import java.util.concurrent.ConcurrentHashMap

@Serializable(with = CosmeticSlot.Serializer::class)
data class CosmeticSlot private constructor(val id: String) {
    companion object {
        private val values = mutableListOf<CosmeticSlot>()
        private val entries = ConcurrentHashMap<String, CosmeticSlot>()

        fun of(id: String): CosmeticSlot {
            return entries.computeIfAbsent(id, ::CosmeticSlot)
        }

        private fun make(id: String): CosmeticSlot {
            return of(id).also { values.add(it) }
        }

        @JvmStatic
        fun values(): List<CosmeticSlot> = values

        @JvmField val BACK = make("BACK")
        @JvmField val EARS = make("EARS")
        @JvmField val FACE = make("FACE")
        @JvmField val FULL_BODY = make("FULL_BODY")
        @JvmField val HAT = make("HAT")
        @JvmField val PET = make("PET")
        @JvmField val TAIL = make("TAIL")
        @JvmField val ARMS = make("ARMS")
        @JvmField val SHOULDERS = make("SHOULDERS")
        @JvmField val SUITS = make("SUITS")
        @JvmField val SHOES = make("SHOES")
        @JvmField val PANTS = make("PANTS")
        @JvmField val WINGS = make("WINGS")
        @JvmField val EFFECT = make("EFFECT")
        @JvmField val CAPE = make("CAPE")
        @JvmField val EMOTE = make("EMOTE")
        @JvmField val ICON = make("ICON")
        @JvmField val TOP = make("TOP")
        @JvmField val ACCESSORY = make("ACCESSORY")
        @JvmField val HEAD = make("HEAD")
    }

    internal object Serializer : KSerializer<CosmeticSlot> {
        private val inner = String.serializer()
        override val descriptor: SerialDescriptor = inner.descriptor
        override fun deserialize(decoder: Decoder) = of(decoder.decodeSerializableValue(inner))
        override fun serialize(encoder: Encoder, value: CosmeticSlot) = encoder.encodeSerializableValue(inner, value.id)
    }
}
